---
title: "Differential expression analysis of NF1 MPNSTs compared to pNFs"
author: "Sara Gosline"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

The goal of this markdown is to describe the differential expression of MPNST samples compared to other NF1 tumors. We are using the patient tumor data only from a diverse number of datasets described in [this Synapse Table](https://www.synapse.org/#!Synapse:syn21221980/tables/).


# Retrieve data

The data is stored in 4 different synapse tables so we need to query and pull down those datasets and join them together. This comprises all the human-related RNA-seq data in the NF data portal.

```{r Get Data, warning=FALSE,echo=FALSE}
library(reticulate)
#get the tidied data with annotations
synapse <- import_from_path("synapseclient",path='/usr/local/lib/python3.7/site-packages/')
syn <- synapse$Synapse()
syn$login()
library(tidyverse)

#tables for expressed genes
tabids<-syn$tableQuery('select distinct tableId from syn21221980')$filepath%>%
  readr::read_csv()
vars="specimenID,individualID,Symbol,totalCounts,zScore,tumorType,nf1Genotype,sex,isCellLine,studyName,transplantationType"

full.tab<-do.call(rbind,lapply(tabids$tableId,function(x) syn$tableQuery(paste('select',vars,'from',x,"where diagnosis=\'Neurofibromatosis 1\'"))$filepath%>%
  readr::read_csv()))

cl=which(full.tab$isCellLine=='TRUE')
if(length(cl)>0)
  full.tab<-full.tab[-cl,]

xe=which(full.tab$transplantationType=='xenograft')
if(length(xe)>0)
  full.tab<-full.tab[-xe,]

full.tab$tumorType<-sapply(full.tab$tumorType,function(x) gsub('Malignant peripheral nerve sheath tumor','Malignant Peripheral Nerve Sheath Tumor',x))

full.tab<-full.tab%>%subset(tumorType%in%c('Malignant Peripheral Nerve Sheath Tumor','Plexiform Neurofibroma'))

DT::datatable(full.tab%>%group_by(tumorType,studyName)%>%summarize(`Number of Samples`=n_distinct(specimenID)))

```

Now that we have the data we can use the counts to calculate differential expression between MPNST and other tumors.

# Calculate Differential Expression

We use DESeq2 to calculate differential expression between the 20 MPNST samples and the other NF1-related tumors.

```{r Calculate Diffex,warning=FALSE, echo=FALSE}
#calculate diffex

combined.mat=round(reshape2::acast(full.tab,Symbol~specimenID,value.var="totalCounts",fun.aggregate=function(x) mean(x,na.rm=TRUE)))
missing=which(apply(combined.mat,1,function(x) any(is.na(x))))
if(length(missing)>0)
  combined.mat=combined.mat[-missing,]

###metadata
vars<-full.tab%>%dplyr::select(specimenID,sex,tumorType,studyName)%>%
  unique()%>%
  remove_rownames()%>%
  column_to_rownames('specimenID')

#vars$isCellLine=tolower(vars$isCellLine)
vars$tumorType<-make.names(vars$tumorType)

expr<-which(apply(combined.mat,1,function(x) any(x>1)))
if(length(expr)>0)
  combined.mat<-combined.mat[expr,]
##run DESeq2
library(DESeq2)
dds <- DESeqDataSetFromMatrix(countData = combined.mat,
  colData = vars[colnames(combined.mat),],
  design= ~ 0+tumorType+sex)

dds <- DESeq(dds)

resultsNames(dds)

all.tt<-resultsNames(dds)[grep('tumorType',resultsNames(dds))]
sampls=c('tumorTypeMalignant.Peripheral.Nerve.Sheath.Tumor')
res=results(dds,
    contrast=list(sampls,setdiff(all.tt,sampls)),
    listValues=c(1/length(sampls),-1/length(setdiff(all.tt,sampls))))


#filter for significant genes
res.sig<-subset(res,padj<0.01)%>%subset(abs(log2FoldChange)>1)

#show table
DT::datatable(data.frame(res.sig))
```

We filter with those genes that exhibit a log2 fold change of _over_  and an adjusted p.value < 0.01 to get `r nrow(res.sig)` genes.

# Functional Enrichment

Next step is to use functional enrichment tools to compare these genes to various gene sets (GO, KEGG, REACTOME) to see what is showing up in these genes  that is of interest. This evaluates the genes in the table above.

```{r Functional Enrichment, warning=FALSE, echo=FALSE}

library(clusterProfiler)
library(biomaRt)
mart <- useMart('ensembl',dataset='hsapiens_gene_ensembl')

  all.genes<-data.frame(res)%>%rownames_to_column('gene')%>%dplyr::select(gene,log2FoldChange,padj)

    #do kegg enrichment
    entrez_list <- getBM(filters = "hgnc_symbol",
      attributes = c("hgnc_symbol", "entrezgene_id"),
      values = all.genes$gene, mart = mart)%>%
        rename(hgnc_symbol='gene')%>%
        left_join(all.genes,by='gene')

    #all genes, for GSEA
    all.eg=entrez_list$log2FoldChange%>%set_names(entrez_list$entrezgene_id)

    #significant genes

    sig.eg=entrez_list%>%
      subset(padj<0.01)%>%
      subset(abs(log2FoldChange)>2)
    eg<-sig.eg$log2FoldChange%>%set_names(sig.eg$entrezgene_id)

```

## KEGG enrichment

Here we calculate the enrichment in KEGG terms. These are primarily pathways.

```{r Kegg Enrichment, warning=FALSE, echo=FALSE}
    kk <- clusterProfiler::enrichKEGG(gene = names(sort(abs(eg),decreasing=T)), organism = "hsa",
      pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.1)
    full.path<-summary(kk)%>%dplyr::select(ID,Description,GeneRatio,pvalue,p.adjust,qvalue,geneID)%>%
      mutate(genelist=paste(sapply(unlist(strsplit(geneID,split='/')),function(x) entrez_list[match(x,entrez_list$entrezgene_id),'gene']),collapse=','))%>%
      dplyr::select(-geneID)

  heatplot(kk, showCategory=10, foldChange = 2^eg)

    #post ranked pathways to gdb pathways
  dotplot(kk,showCategory=20)
```

There is not much there! Maybe we are too stringent?

## GO Enrichment

Now we can try GO enrichment. We focus on the 'biological process' ontology as we are more interested in those terms.

```{r GO Enrichment, warning=FALSE, echo=FALSE}
    kk <- clusterProfiler::enrichGO(gene = names(sort(abs(eg),decreasing=T)), OrgDb='org.Hs.eg.db',ont='BP',
      pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.1)

    full.path<-summary(kk)%>%dplyr::select(ID,Description,GeneRatio,pvalue,p.adjust,qvalue,geneID)%>%
      mutate(genelist=paste(sapply(unlist(strsplit(geneID,split='/')),function(x) entrez_list[match(x,entrez_list$entrezgene_id),'gene']),collapse=','))%>%
      dplyr::select(-geneID)

  heatplot(kk, showCategory=10, foldChange = 2^eg)

    #post ranked pathways to gdb pathways
#  heatplo(kk,showCategory=20)
  ##now we try GO enrichment
    DT::datatable(full.path)


```
Most of the pathways are transcriptional and/or epigenetic. We will now move to other modes of analysis. Perhaps Reactome will have more metabolic signatures.

## Reactome Enrichment

Now we search for enrichment using the Reactome library.I know very little about this library but thought it'd be worth a try.

```{r reactome enrichment,warning=FALSE,echo=FALSE}
  require(ReactomePA)
  kk <- enrichPathway(gene = names(sort(abs(eg),decreasing=T)), pAdjustMethod = "BH", pvalueCutoff = 0.05, qvalueCutoff = 0.1)

    full.path<-as.data.frame(kk)%>%dplyr::select(ID,Description,GeneRatio,pvalue,p.adjust,qvalue,geneID)%>%
      mutate(genelist=paste(sapply(unlist(strsplit(geneID,split='/')),function(x) entrez_list[match(x,entrez_list$entrezgene_id),'gene']),collapse=','))%>%
      dplyr::select(-geneID)

  cnetplot(kk, showCategory=10,categorySize = "pvalue", foldChange = 2^eg)

    #post ranked pathways to gdb pathways
  heatplot(kk,showCategory=20, foldChange = 2^eg)
  DT::datatable(full.path)
```

# GSEA enrichment with MSIGDB

Lastly we can look for enrichment of all the genes with MSIGDB, which is a database of explicit expression profiles. This uses a slightly different enrichment signature which evaluates the entire list of genes to look at patterns of genes that are universally up-regulated or down-regulated.

## Full analysis

The gene set I used is the [C2 Gene list](http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=C2) which comprises other expression-based perturbations

```{r GSEA analysis,warning=FALSE,echo=FALSE}

##get gene list

gmtfile <- system.file("extdata", "wikipathways-20180810-gmt-Homo_sapiens.gmt", package="clusterProfiler")
gmtfile<-'c2.all.v7.0.entrez.gmt'
c5 <- read.gmt(gmtfile)

kk <- clusterProfiler::GSEA(geneList=sort(all.eg,decreasing=T), TERM2GENE=c5)

full.path<-as.data.frame(kk)%>%dplyr::select(ID,Description,enrichmentScore,pvalue,p.adjust,qvalues)#%>%
#      mutate(genelist=paste(sapply(unlist(strsplit(core_enrichment,split='/'))[1:10],function(x) entrez_list[match(x,entrez_list$entrezgene_id),'gene']),collapse=','))%>%
 #     dplyr::select(-core_enrichment)

    #post ranked pathways to gdb pathways
  heatplot(kk,showCategory=20, foldChange = 2^eg)
  DT::datatable(full.path)


```

## Focused analysis

We found some metabolism related pathways here, shown below. These plots represent the GSEA enrichment statistic. Genes are ranked by differential expression (along x-axis), if the gene is in the gene set there is a hash mark along the 0.0 point. The green line represents the running score. The red dashed line represents the point at which the statistic is called significant.


```{r metab, warning=FALSE,echo=FALSE}

mets<-full.path$ID[c(grep('METAB',full.path$ID),grep("GLUT",full.path$ID))]

res=lapply(mets,function(x) 
  clusterProfiler::gseaplot(kk, geneSetID = x,by='runningScore', title = x)
)
cowplot::plot_grid(plotlist=res[1:2],ncol=1)
cowplot::plot_grid(plotlist=res[3:4],ncol=1)
cowplot::plot_grid(plotlist=res[5:6],ncol=1)
cowplot::plot_grid(plotlist=res[7:8],ncol=1)
cowplot::plot_grid(plotlist=res[9:10],ncol=1)
cowplot::plot_grid(plotlist=res[11:12],ncol=1)
cowplot::plot_grid(plotlist=res[13],ncol=1)

#cowplot::plot_grid(res)
```

